File: tethers-0.1/portable-rust/wrappers/go/tethers.go
package tethers

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"os/exec"
	"time"
)

type Decision string
const ( Allow Decision = "ALLOW"; Ask Decision = "ASK"; Deny Decision = "DENY" )
type Result struct { Decision Decision `json:"decision"`; Rule string `json:"matched_rule"`; Reason string `json:"reason"`; Error string `json:"error"` }

func Evaluate(binary string, request []byte, timeout time.Duration) Result {
	ctx, cancel := context.WithTimeout(context.Background(), timeout); defer cancel()
	cmd := exec.CommandContext(ctx, binary, "evaluate"); cmd.Stdin = bytes.NewReader(request)
	out, err := cmd.Output()
	if ctx.Err() != nil { return Result{Decision: Deny, Error: "Tethers evaluation timed out"} }
	if err != nil { return Result{Decision: Deny, Error: fmt.Sprintf("Tethers failed: %v", err)} }
	return decode(out)
}

func decode(out []byte) Result {
	var result Result
	if err := json.Unmarshal(out, &result); err != nil { return Result{Decision: Deny, Error: "invalid Tethers response"} }
	if result.Decision != Allow && result.Decision != Ask && result.Decision != Deny { return Result{Decision: Deny, Error: "unknown Tethers decision"} }
	return result
}

File: tethers-0.1/portable-rust/wrappers/python/tethers.py
"""Tiny subprocess adapter for the canonical Tethers JSON protocol."""
import json
import subprocess
from typing import Any

DECISIONS = {"ALLOW", "ASK", "DENY"}

def decode_response(stdout: str) -> dict[str, Any]:
    try:
        result = json.loads(stdout)
    except (TypeError, json.JSONDecodeError) as exc:
        return {"decision": "DENY", "error": f"invalid Tethers response: {exc}"}
    if result.get("schema_version") != "1":
        return {"decision": "DENY", "error": "Tethers response schema mismatch"}
    if result.get("decision") not in DECISIONS:
        return {"decision": "DENY", "error": "unknown Tethers decision"}
    return result

def evaluate(binary: str, request: dict[str, Any], timeout: float = 5.0) -> dict[str, Any]:
    try:
        completed = subprocess.run([binary, "evaluate"], input=json.dumps(request), text=True,
                                   capture_output=True, timeout=timeout, check=False)
    except (OSError, subprocess.TimeoutExpired) as exc:
        return {"decision": "DENY", "error": f"Tethers unavailable: {exc}"}
    if completed.returncode != 0:
        return {"decision": "DENY", "error": f"Tethers exited with {completed.returncode}"}
    return decode_response(completed.stdout)
